workflow/controller/operator.go

```diff
@@ -1117,6 +1124,18 @@ func (woc *wfOperationCtx) podReconciliation(ctx context.Context) error {
 		node, err := woc.wf.Status.Nodes.Get(nodeID)
 		if err == nil {
 			if newState := woc.assessNodeStatus(pod, node); newState != nil {
+				if newState.Succeeded() {
+					tmpl, err := woc.GetNodeTemplate(node)
+					if err != nil {
+						return
+					}
+					// Check whether the node has output and whether its taskresult is in an incompleted state.
+					if tmpl.HasOutputs() && woc.wf.Status.IsTaskResultIncomplete(node.ID) {
```

Comment:
I'm wondering if it's necessary to include `tmpl.HasOutputs()` in the if statement?

Reply:
I think it is necessary. We can't judge a node IsTaskResultIncomplete only by "woc.wf.Status.IsTaskResultIncomplete", If there no key in TaskResultsCompletionStatus, Maybe the taskresult has not been received yet, or maybe the taskresult is not needed at all.

Reply:
maybe im mossong something, but if `HasOutputs()` evaluates to true, it could still be possible that outputs haven't been initialized and therefore no key exists in `TaskResultCompletionStatus`. Does this actually give us more protection? This is only here for when task results are incomplete, so it depends on keys being in `TaskResultsCompletionStatus` anyway.

Reply:
As far as I understand, every TaskResult that's received will be marked [incomplete](https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/taskresult.go#L72) and then marked [complete](https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/taskresult.go#L69), regardless of whether there are outputs defined, right? And it looks like `IsTaskResultIncomplete()` will return true if there's no entry in the map for the node yet. So, it seems like if we could just test for that, no?

Reply:
Yes, I overlooked something. Some tasks do not define output. We also created taskresult. Can we add some restrictions? For tasks without output, cancel
```
wfExecutor.InitializeOutput(bgCtx)
defer wfExecutor.HandleError(bgCtx), 
```
because after all, this increases the pressure on the apiserver.

Reply:
So, 2 fewer API calls per Pod that doesn't have "Outputs"? I can see that. I do, on the other hand, kind of like the simplicity of just doing it across the board. And also, that label essentially means that all outputs of any kind are done, including [Logs](https://github.com/argoproj/argo-workflows/blob/main/cmd/argoexec/commands/wait.go#L66).

Reply:
Actually, we probably need to keep it the way it is - `checkReconciliationComplete()` will verify that all TaskResults of any kind are completely processed before allowing Workflow to be marked completed ([here](https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/operator.go#L2319))

Reply:
> So, 2 fewer API calls per Pod that doesn't have "Outputs"? I can see that. I do, on the other hand, kind of like the simplicity of just doing it across the board and not having to have special if statements on the Controller side. And also, that label essentially means that all outputs of any kind are done, including [Logs](https://github.com/argoproj/argo-workflows/blob/main/cmd/argoexec/commands/wait.go#L66).

OK, If each node has taskresult. I  will change it.